app/drivers/zephyr/battery_voltage_divider.c

```diff
@@ -0,0 +1,214 @@
+        LOG_DBG("ADC raw %d ~ %d mV => %d mV\n", drv_data->adc_raw, val, millivolts);
+        uint8_t percent = lithium_ion_mv_to_pct(millivolts);
+        LOG_DBG("Percent: %d", percent);
+
+        drv_data->voltage = millivolts;
+        drv_data->state_of_charge = percent;
+    } else {
+        LOG_DBG("Failed to read ADC: %d", rc);
+    }
+
+    // Disable power GPIO if present
+    if (drv_data->gpio) {
+        rc = gpio_pin_set(drv_data->gpio, drv_cfg->power_gpios.pin, 0);
+
+        if (rc != 0) {
+            LOG_DBG("Failed to disable ADC power GPIO: %d", rc);
+        }
+    }
+
+    return rc;
+}
+
+static int bvd_channel_get(struct device *dev, enum sensor_channel chan, struct sensor_value *val) {
+    struct bvd_data *drv_data = dev->driver_data;
+
+    switch (chan) {
+    case SENSOR_CHAN_GAUGE_VOLTAGE:
+        val->val1 = drv_data->voltage / 1000;
+        val->val2 = (drv_data->voltage % 1000) * 1000U;
+        break;
+
+    case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+        val->val1 = drv_data->state_of_charge;
+        val->val2 = 0;
+        break;
+
+    default:
+        return -ENOTSUP;
+    }
+
+    return 0;
+}
+
+static const struct sensor_driver_api bvd_api = {
+    .sample_fetch = bvd_sample_fetch,
+    .channel_get = bvd_channel_get,
+};
+
+static int bvd_init(struct device *dev) {
+    struct bvd_data *drv_data = dev->driver_data;
+    const struct bvd_config *drv_cfg = dev->config_info;
+
+    drv_data->adc = device_get_binding(drv_cfg->io_channel.label);
+
+    if (drv_data->adc == NULL) {
+        LOG_ERR("ADC %s failed to retrieve", drv_cfg->io_channel.label);
+        return -ENOENT;
+    }
+
+    int rc = 0;
+
+    if (drv_cfg->power_gpios.label) {
+        drv_data->gpio = device_get_binding(drv_cfg->power_gpios.label);
+        if (drv_data->gpio == NULL) {
+            LOG_ERR("Failed to get GPIO %s", drv_cfg->power_gpios.label);
+            return -ENOENT;
+        }
+        rc = gpio_pin_configure(drv_data->gpio, drv_cfg->power_gpios.pin,
+                                GPIO_OUTPUT_INACTIVE | drv_cfg->power_gpios.flags);
+        if (rc != 0) {
+            LOG_ERR("Failed to control feed %s.%u: %d", drv_cfg->power_gpios.label,
+                    drv_cfg->power_gpios.pin, rc);
+            return rc;
+        }
+    }
+
+    drv_data->as = (struct adc_sequence){
+        .channels = BIT(0),
+        .buffer = &drv_data->adc_raw,
+        .buffer_size = sizeof(drv_data->adc_raw),
+        .oversampling = 4,
+        .calibrate = true,
+    };
+
+#ifdef CONFIG_ADC_NRFX_SAADC
+    drv_data->acc = (struct adc_channel_cfg){
+        .gain = ADC_GAIN_1_5,
+        .reference = ADC_REF_INTERNAL,
+        .acquisition_time = ADC_ACQ_TIME(ADC_ACQ_TIME_MICROSECONDS, 40),
+        .input_positive = SAADC_CH_PSELP_PSELP_AnalogInput0 + drv_cfg->io_channel.channel,
+    };
+
+    drv_data->as.resolution = 12;
+#else
+#error Unsupported ADC
```

Comment:
So the whole `battery_voltage_divider` is dependend on `ADC_NRFX_SAADC`?
I don't understand why, it sounds like just having an ADC is enough, is there a specific functionality you rely on that I didn't get?

Reply:
So the reason for this is that the ADC channel config set up really depends on the chip we're using as well as the voltage divider configuration from my understanding. For now almost everyone is using nrf52 with a very specific voltage divider setup recommended from Nordic (hence the specific gain, reference, acquisition time, etc). The reason it's set up this way is so we can add more else ifs in the future to handle other chips. Possibly making the gain/ref/acquisition options in the DT, but I'm not certain of this, and at this point we're all using the same set up.

Reply:
Thanks for the explanation but I still do not get it:
The voltage divider comes from the device tree and is not dependent on the ADC itself. As for the ADC configuration, this should be generic. If another MCU is used, I would expect that the ADC reading would work out of the box with the configuration you set up. So I would just remove this `#ifdef` pattern.
But I may be missing something, so if you find it simpler/better to keep as is, all good :)

Reply:
Removing the `ifdef `pattern would result in errors for non-nrf ADCs because I reference `SAADC_CH_PSELP_PSELP_AnalogInput0`, which is only for nrfx. This error is more explicit to someone who is trying to use a new MCU (with different ADC reqs).

Along with this, from my understanding other ADCs need different parts of the adc_channel_cfg defined compared to what nrf ADCs need. I'm personally not interested in researching every other MCU Zephyr supports to make a bunch of logic to handle each type. At the moment we only use nrf52, and every single one uses these specific gains, references, and acquisition times. A generic solution would be a nice enhancement to this in the future, but I don't think it's necessary to have this driver added to ZMK.

I personally couldn't find any generic samples that supported any ADC in Zephyr's repo. I believe the reasons for that are the same reasons I haven't created a generic solution. I could definitely be wrong about some of this though. I'm basing it off of what samples I could find and the source code.

Reply:
Oh yeah sorry I missed the `SAADC_CH_PSELP_PSELP_AnalogInput0` and indeed I see that the examples in zephyr repo use MCU specific gain/ref/acqu_time config.